### ISSUE TITLE
[Bug]: Babel/ts can't parse `type tmp = number?;` (tsc can)

### ISSUE THREAD

**Issue by kungfooman** [babel/babel#16073]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

[REPL](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=C4TwDgpgBMC2ZQLxQHYFdYCMICcD8A3EA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=typescript&prettier=false&targets=&version=7.23.2&externalPlugins=&assumptions=%7B%7D):

![image](https://github.com/babel/babel/assets/5236548/d3ccdbf4-0fdc-4b5d-8d0b-f33d1082a2a7)

[TS AST Viewer](https://ts-ast-viewer.com/#code/C4TwDgpgBMC2ZQLxQHYFdYCMICcD8A3EA):

![image](https://github.com/babel/babel/assets/5236548/23f4db25-42c4-4366-881f-3b6fe9c38130)

```js
import {parse} from '@babel/parser';
parse('type tmp = number?;', {plugins: ['typescript']});
```

![image](https://github.com/babel/babel/assets/5236548/0cde9a30-c953-4452-a46b-dce598238a52)

### Current and expected behavior

Current behaviour: parse error
Expected behaviour: parse it like tsc does

### Environment

```js
"@babel/parser": "^7.23.0"
```

### Possible solution

Fix parser by adding two node types:
 - JSDocNullableType
 - JSDocNonNullableType (replace ? with !)

![image](https://user-images.githubusercontent.com/6856943/39607069-58d0b3d2-4ee5-11e8-8d0e-ec680c48e238.png)

### Additional context

@nicolo-ribaudo mentioned the need for these kinds of parse issues here: https://github.com/babel/babel/issues/16048#issuecomment-1763556413

> It's ok if Babel does not error on every invalid TS code (since the type checker will error anyway), but we must parse every TS code that tsc can parse.

**Comment by liuxingbaoyu** [comment]
https://www.typescriptlang.org/play?#code/C4TwDgpgBMC2ZQLxQHYFdYCMICcD8A3EA
This is not valid for tsc.

**Comment by kungfooman** [comment]
@liuxingbaoyu Please read what you have marked yourself with thumbs up:

> It's ok if Babel does not error on every invalid TS code (since the type checker will error anyway), but we must parse every TS code that tsc can parse.

_Originally posted by @nicolo-ribaudo in https://github.com/babel/babel/issues/16048#issuecomment-1763556413_

...we *must* parse... what the type checker claims afterwards doesn't matter (hence look at the AST level)

**Comment by liuxingbaoyu** [comment]
According to my understanding, this means that Babel may parse illegal ts code, and Babel may not necessarily parse the code with tsc errors.

**Comment by kungfooman** [comment]
> this means that Babel may parse illegal ts code

No, it's not illegal, it's as legal as it gets - implemented by tsc itself. You objected yourself to my view that maybe we shouldn't consider `tsc` as "holy truth" and now you want to derive from that view yourself.

Type checking errors you see in TS PlayGround / Monaco play no role here. You can easily find examples that Babel-TS happily parses:

```js
const str = 'type string = 0;'
parse(str, {plugins: ['typescript']});
```

![image](https://github.com/babel/babel/assets/5236548/7ac70a96-9fec-4c4a-8ce6-ea12302a2533)

https://www.typescriptlang.org/play?#code/C4TwDgpgBAzsBOBLAdgcygXigBgNxA

![image](https://github.com/babel/babel/assets/5236548/2e3ee1f3-3256-4142-8150-782f7d633e46)

Oops? But this doesn't matter because AST is AST. We can't argue X in issue A and opposite of that in issue B...

**Comment by liuxingbaoyu** [comment]
I think Babel parser only officially supports parsing code that doesn't have parse errors/type errors in tsc.
tsc can parse it for better error prompts during development, but Babel does not need to do this. Babel is expected to accept TS code without any errors in tsc.
Also if you look at it in typescript-eslint, you will find that this AST is not part of the TS, it is a node in jsdoc, and babel does not parse jsdoc.
https://typescript-eslint.io/play#ts=4.6.2&showAST=es&fileType=.tsx&code=LAKALgngDgpgBGAtlOBeOA7ArogRjAJwH4BuIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

**Comment by nicolo-ribaudo** [comment]
TypeScript does indeed parse that code and report a "syntax error" (the error message is `'?' at the end of a type is not valid TypeScript syntax`) at type-checking time. Normally, way to tell whether something in TS is a syntax error is that it's reported in `parseDiagnostics` in their AST. This specific error is a weird combination (it claims to be one thing, it is reported as the other), and I lean towards considering it as a syntax error and not as a type error.

If there are examples of real-world projects using the `number?` syntax by suppressing the error with `@ts-expect-error`/`@ts-ignore`, I'm happy to change my stance.

@kungfooman Your message comes off as aggressive. It gives the vibe of "you disagreed with me in the other issue, so now I'm proving that you are wrong". Please keep this discussion constructive.

**Comment by kungfooman** [comment]
@nicolo-ribaudo Yes, I realize that this case is a weird combination, it claims to be an error while having a perfect and meaningful AST without a report in `parseDiagnostics`:

```js
const str = `type tmp = number?;`;
const ret = ts.createSourceFile('repl.ts', str, ts.ScriptTarget.Latest, true);
ret.parseDiagnostics; // empty array
```

That puts it on the same level as e.g. `type string = 0;` and is fully in line with your own previous view:

> It's ok if Babel does not error on every invalid TS code (since the type checker will error anyway), but we must parse every TS code that tsc can parse.

> If there are examples of real-world projects using the number? syntax by suppressing the error with @ts-expect-error/@ts-ignore, I'm happy to change my stance.

Yes, I'm the author of a real-world project and I require this. I'm parsing JSDoc comments for https://runtimetypeinspector.org/ currently via `tsc`, but otherwise I use Babel for everything else. However `tsc` is a large and complicated dependency and I would rather *only* use `Babel`. But right now I can't, simply because Babel-TS doesn't match the abilities of `tsc`.

It also comes across as aggressive when people "suddenly change their mind" because it just happens to be convenient, just saying. Despite that, I hope we can look over that now and try to find a solution to make Babel-TS just as capable as `tsc`.

What do you think about this as possible compromise?

```js
const str = 'number?';
parse(str, {plugins: ['typescript-jsdoc']});
```

`tsc` doesn't really make that separation, but if we follow your syntax error logic, then we require some other way of doing what `tsc` does for JSDoc type AST's.

**Comment by kungfooman** [comment]
Here is another one that `tsc` can parse, but Babel-TS can't:

```js
const str = `type tmp = Array<>;`;
const ret = ts.createSourceFile('repl.ts', str, ts.ScriptTarget.Latest, true);
console.log("parseDiagnostics", ret.parseDiagnostics);
ret.getText();
```

Output:

![image](https://github.com/babel/babel/assets/5236548/13e62737-0c3e-4093-b157-4efbe0b9a6e6)

Babel-TS:

```js
const str = `type tmp = Array<>;`;
parse(str, {plugins: ["typescript"]});
```

![image](https://github.com/babel/babel/assets/5236548/908acf4b-c1ca-452c-a094-09a6b54d9700)

**Comment by liuxingbaoyu** [comment]
> Will Babel ever be able to do what TypeScript can in this regard?

It's probably not possible, supporting JSDOC is too complex for us.
